broker/src/main/java/io/zeebe/broker/system/partitions/impl/StateControllerImpl.java

```diff
@@ -0,0 +1,312 @@
+/*
+ * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH under
+ * one or more contributor license agreements. See the NOTICE file distributed
+ * with this work for additional information regarding copyright ownership.
+ * Licensed under the Zeebe Community License 1.0. You may not use this file
+ * except in compliance with the Zeebe Community License 1.0.
+ */
+package io.zeebe.broker.system.partitions.impl;
+
+import io.atomix.raft.snapshot.PersistedSnapshot;
+import io.atomix.raft.snapshot.PersistedSnapshotListener;
+import io.atomix.raft.snapshot.PersistedSnapshotStore;
+import io.atomix.raft.snapshot.ReceivedSnapshot;
+import io.atomix.raft.snapshot.SnapshotChunk;
+import io.atomix.raft.snapshot.TransientSnapshot;
+import io.atomix.utils.time.WallClockTimestamp;
+import io.zeebe.broker.system.partitions.AtomixRecordEntrySupplier;
+import io.zeebe.broker.system.partitions.SnapshotReplication;
+import io.zeebe.broker.system.partitions.StateController;
+import io.zeebe.db.ZeebeDb;
+import io.zeebe.db.ZeebeDbFactory;
+import io.zeebe.logstreams.impl.Loggers;
+import io.zeebe.util.FileUtil;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.Map;
+import java.util.Optional;
+import java.util.function.ToLongFunction;
+import org.agrona.collections.Object2NullableObjectHashMap;
+import org.slf4j.Logger;
+
+/** Controls how snapshot/recovery operations are performed */
+public class StateControllerImpl implements StateController, PersistedSnapshotListener {
+
+  private static final ReplicationContext INVALID_SNAPSHOT = new ReplicationContext(-1, null);
+  private static final Logger LOG = Loggers.SNAPSHOT_LOGGER;
+
+  private final SnapshotReplication replication;
+  private final Map<String, ReplicationContext> receivedSnapshots =
```

Comment:
Single threaded or need for synchronization?

Reply:
It looks like there might be a memory leak here. 
- entries which are marked as INVALID_SNAPSHOT will never be evicted from the map
- entries which never receive the last chunk will never be evicted from the map.

I think it would make sense to evict entries more aggressively, either by looking at the start time of the replication context or by looking at the index of the snapshot and removing all entries of snapshots which are older than the last successfully replicated snapshot.

Reply:
Ok lets do an follow up on this since this was also the case before the refactoring. I created an issue for it https://github.com/zeebe-io/zeebe/issues/4671